### DISCUSSION TITLE
Conditional Dirichlet BCs

### DISCUSSION THREAD

**Discussion by reverendbedford** [idaholab/moose#20563] [Q&A General] [answered]
Is there a way to apply "conditional" Dirichlet BCs to part of a side set?  For example, say I have a simulation where I only want to apply a Dirichlet condition over a location that changes as a function time (but will always be contained within a sideset on the mesh).

For a Neumann BC this type of thing is possible: just use a FunctionNeumannBC where you apply whatever BC you want in the current/active region and zero outside of that.  The zero here won't affect the problem.

But this isn't so simple for DirichletBCs, as outside the active region I want to impose no constraint on the problem, not a zero boundary condition.

The background here is that I have a simulation where I'm trying to effectively do rigid contact but using a stiff contact surface is proving to be numerically quite difficult...

**Comment by hugary1995** [comment, marked as answer]
There is `ElementSubdomainModifier` that modifies element subdomain during the simulation. I wonder if you could follow what's in there and do something similar to sidesets/nodesets. It's going to be a custom user object.

**Reply by reverendbedford** [reply]
I guess the hacky way would be to subclass [PenaltyDirichletBC](https://mooseframework.inl.gov/source/bcs/PenaltyDirichletBC.html) to take a function for the penalty term.  But I suspect that's going to put me right back in convergence purgatory.  Still, I'm desperate so maybe I'll see what happens.

**Reply by hugary1995** [reply]
Let me know if it doesn't work. I can write a `SidesetModifier`.

**Reply by reverendbedford** [reply]
Somewhat to my surprise this worked.  So the idea is to impose the Dirichlet BC with a penalty, but then use some sort of step function for the penalty stiffness so that it's only non-zero in the region you want to impose the BC.  This seems like a very specialized type of thing to do but if there's any interest in having it in the framework I'd be happy to share the code.

**Reply by lyyc199586** [reply]
Hi Mark @reverendbedford, I meet almost the same situation here (Dirichlet BC on a time-varying boundary), can you share the conditional BC you wrote?

**Reply by hugary1995** [reply]
https://github.com/Argonne-National-Laboratory/deer/blob/master/src/bcs/QuasiContactBC.C

**Reply by lyyc199586** [reply]
👍
